### DISCUSSION TITLE
Add loading state for atomWithStorage

### DISCUSSION THREAD

**Discussion by christiearcus** [pmndrs/jotai#2369] [Ideas]
I'm having a bit of trouble using `atomWithStorage`. It doesn't load synchronously as the first render is the initial value before re-rendering with the value from local storage. Because of this it's a bit hard to work with, and it'd be great to have a loading state so I know when it's ready to be used as with local storage there could be a value or undefined. It's be great to not just have the initial value persist. 

Could you let me know if there's a workaround for this?

**Comment by dai-shi** [comment]
I think we have a similar discussion before.
Do you need Suspense for the loading state?
Otherwise, it should be possible with custom storage.

Take a very simple example:
```js
const storage = {
  ...localStorage,
  getItem(key) {
    return localStorage.getItem(key) ?? "EMPTY VALUE"
  },
}

const anAtom = atomWithStorage("foo", "DEFAULT VALUE", storage);
```

If we need JSON:
```js
const EMPTY_DATA = { /* ... */ }
const DEFAULT_DATA = { /* ... */ }
const baseStorage = createJSONStorage();
const storage = {
  ...baseStorage,
  getItem(key) {
    return baseStorage.getItem(key, EMPTY_DATA)
  },
}

const anAtom = atomWithStorage('foo', DEFAULT_DATA, storage)
```

It might be worth considering adding `emptyData` option.
```js
// an idea
const anAtom = atomWithStorage('foo', DEFAULT_DATA, undefined, { emptyData: EMPTY_DATA })
```

**Comment by tancop** [comment]
Call it with `{ getOnInit: true }` as the fourth parameter like this:
```typescript
const fooAtom = atomWithStorage('foo', 'bar', localStorage, { getOnInit: true });
```
This will load the value from storage when the atom gets initialized so components never see the initial value.

**Reply by Falven** [reply]
This doesn't work with hydration. Op's question is still relevant. It would be great if the atom simply returned, for example:

```ts
'use client';

import { atomWithStorage, createJSONStorage } from 'jotai/utils';

const storage = createJSONStorage(() => sessionStorage);

export const myAtom = atomWithStorage('myAtom', null, { ...storage });
```

```tsx
const [value, setValue, valueIsLoading] = useAtom(myAtom);

if (valueIsLoading) {
    return <span>Value is Loading</span>
}

// value can be null (e.g. nothing has been set by the user yet, but we have already loaded from storage.)
return <span>Value loaded: {value}</span>
```

rather than forcing the user to introduce some new abstract value of "empty" as @dai-shi mentioned as well. I think [this is the discussion you were referencing](https://github.com/pmndrs/jotai/discussions/1432)

Here's the example with an abstract value of "empty" where undefined means not yet loaded and null means not set.

```ts
'use client';

import { atomWithStorage, createJSONStorage } from 'jotai/utils';

const baseStorage = createJSONStorage<MyType | null | undefined>(() => sessionStorage);

const storage = {
  ...baseStorage,
  // Get the item in storage or "null" when one is not set.
  getItem: (key: string) => baseStorage.getItem(key, null)
};

// Return the value in storage or "undefined" while it's not available.
export const myAtom = atomWithStorage<MyType | null | undefined>('myAtom', undefined, storage);
```

```tsx
const [value, setValue] = useAtom(myAtom);

if (value === undefined) {
    return <span>Value is Loading</span>
}

// value can be null (e.g. nothing has been set by the user yet, but we have already loaded from storage.)
return <span>Value loaded: {value}</span>
```

Specifically, this solves for things like setting some default value that isn't available until render to the atom if none is set in storage.

**Reply by dr-nyt** [reply]
@Falven you are a life saver, this works beautifully, except a big caveat of then having to eternally check for `undefined`. 

My solution was to make the state a part of the data:
```ts
'use client';
import { atomWithStorage, createJSONStorage } from 'jotai/utils';

type MyType = { data: MyDataT | null, initialized: boolean };

const baseStorage = createJSONStorage<MyType>(() => localStorage);
const storage = {
  ...baseStorage,
  // Get the item in storage or default value with `initialized` set to `true`  when one is not set.
  getItem: (key: string) => baseStorage.getItem(key, { data: null, initialized: true })
};

// Return the value in storage or default value with `initialized` set to `false` while it's not available.
export const myAtom = atomWithStorage<MyType>('myAtom', {data: null, initialized: false}, storage);
```

```ts
const [value, setValue] = useAtom(myAtom);

if (!value.initialized) {
    return <span>Value is Loading</span>
}

// value can be null (e.g. nothing has been set by the user yet, but we have already loaded from storage.)
return <span>Value loaded: {value}</span>
```
Why? Because this value is always available except for a few milliseconds in the start, and it makes no sense checking for undefined over and over, or adding that ugly `!` at the end. But of course some caution is required like almost everything in React 😅

**Comment by irg1008** [comment]
```ts
const loadingAtom = atom(true)

const sessionAtom = atomWithStorage<SessionInput | undefined>(
  'session',
  sessionStorage.defaultValue,
  sessionStorage,
  { getOnInit: false },
)

export const useSession = () => {
  const [session, setSession] = useAtom(sessionAtom)
  const [loading, setLoading] = useAtom(loadingAtom)

  sessionAtom.onMount = () => {
    setLoading(false)
  }
```

I find this to be the easiest. You could even use useState for this
